# RiboReadCountByFrame
Process ribORF output for DE analysis

#### R package dependency:
```
GenomicRanges
dplyr
stringr
optparse
data.table
```

### Usage:
```
Rscript /PATH/TO/RiboReadCountByFrame.R -f ${Input_list.txt} -p ${ORF.gtf} -r ${RNA.gtf} -o ${Output_prefix}
```

### Input explaination:
${Input\_list.txt} = a single column text file listing all input sam files, for example:
```
Input_list.txt:
/path/to/input1.sam
/path/to/input2.sam
/path/to/input3.sam
```
${ORF.gtf} = the path to your gtf file generated by ribORF or any other ORF annotation pipeline. If run on multiple samples, should first cat all gtfs to combine them. The script will remove duplicates and rename ORFs.

${RNA.gtf} = the path to your reference gtf file for ribORF or any other ORF annotation pipeline. Used to grep gene/transcript information.

${Output} = the path to your output directory and file name prefix.

### Output explaination:
${Output\_prefix}.gtf: filtered and renamed gtf file, can be load on IGV for visualization.

${Output\_prefix}\_TranscriptInfoSummary.tsv: organized table for gene infomation, extracted from ${RNA.gtf}.

${Output\_prefix}\_SampleID\_GeneCount.tsv: read cound by gene.

${Output\_prefix}\_SampleID\_OrfCountByFrame.tsv: read cound by ORF, seperated by different frames.


### usage example:  
```
Rscript /home/ahe/tools/RiboReadCountByFrame/RiboReadCountByFrame.R \
-f /home/ahe/Analysis/201903_Fleur/data/ribo/191218_out/list.txt \
-p /home/fleur/Analysis/191210_GTFfiles/GBM/GBM_final.gtf \
-r /home/ahe/Analysis/201903_Fleur/data/RNA/191127/GSC_transcript.gtf \
-o ~/Analysis/201903_Fleur/data/ribo/191218_out/RiboReadCountByFrame_out/GSC
```
    
### version descrition:

V1. Wrote by Tony C. 

V2. Updated by Arthur, added mouse genome support. Script renamed.

V3. Updated by Arthur, new output: organized ORF gtf & \_GeneCount.tsv.

V4. Updated by Arthur, now getting gene information from RNA.gtf rather than fixed R file, added capacity to process multiple sam file to save time.

